azkaban-common/src/main/java/azkaban/metric/MetricReportManager.java

```diff
@@ -0,0 +1,220 @@
+ * <p> Metric Manager 'has a' relationship with :-
+ * <ul>
+ * <li>all the metric Azkaban is tracking</li>
+ * <li>all the emitters Azkaban is supposed to report metrics</li>
+ * </ul></p>
+ */
+public class MetricReportManager {
+  /**
+   * Maximum number of metrics reporting threads
+   */
+  private static final int MAX_EMITTER_THREADS = 4;
+  private static final Logger _logger = Logger.getLogger(MetricReportManager.class);
+
+  /**
+   * List of all the metrics that Azkaban is tracking
+   * Manager is not concerned with type of metric as long as it honors IMetric contracts
+   */
+  private List<IMetric<?>> _metrics;
+
+  /**
+   * List of all the emitter listening all the metrics
+   * Manager is not concerned with how emitter is reporting value.
+   * Manager is only responsible to notify all emitters whenever an IMetric wants to be notified
+   */
+  private List<IMetricEmitter> _metricEmitters;
+  private ExecutorService _executorService;
+  // Singleton variable
+  private static volatile MetricReportManager _instance = null;
+  private static boolean _isManagerEnabled;
+
+  private MetricReportManager() {
+    _logger.debug("Instantiating Metric Manager");
+    _executorService = Executors.newFixedThreadPool(MAX_EMITTER_THREADS);
+    _metrics = new ArrayList<IMetric<?>>();
+    _metricEmitters = new LinkedList<IMetricEmitter>();
+    enableManager();
+  }
+
+  /**
+   * @return true, if we have enabled metric manager from Azkaban exec server
+   */
+  public static boolean isAvailable() {
+    return _instance != null && _isManagerEnabled;
+  }
+
+  /**
+   * Get a singleton object for Metric Manager
+   */
+  public static MetricReportManager getInstance() {
+    if (_instance == null) {
+      synchronized (MetricReportManager.class) {
+        if (_instance == null) {
+          _logger.info("Instantiating MetricReportManager");
+          _instance = new MetricReportManager();
+        }
+      }
+    }
+    return _instance;
+  }
+
+  /***
+   * each element of metrics List is responsible to call this method and report metrics
+   * @param metric
+   */
+  public void reportMetric(final IMetric<?> metric) {
+    if (metric != null && isAvailable()) {
+      try {
+        final IMetric<?> metricSnapshot;
+        // take snapshot
+        synchronized (metric) {
+          metricSnapshot = metric.getSnapshot();
+        }
+        _logger.debug(String.format("Submitting %s metric for metric emission pool", metricSnapshot.getName()));
+        // report to all emitters
+        for (final IMetricEmitter metricEmitter : _metricEmitters) {
+          _executorService.submit(new Runnable() {
+            @Override
+            public void run() {
+              try {
+                metricEmitter.reportMetric(metricSnapshot);
+              } catch (Exception ex) {
+                _logger.error(String.format("Failed to report %s metric due to %s", metricSnapshot.getName(),
+                    ex.toString()));
```

Comment:
It is better not to call toString() method.  Let logger.error prints out the entire stack trace.